Add custom metric filters for common scaling errors to default cloudwatch dashboard

HWong-sudo opened the pull request:
### Description of changes
* Improved visibility of known errors from ParallelCluster daemons by adding custom metricFilters for specified common errors 
* Added graph widgets to visualize new metrics as shown in the image below
* Added configuration template option to enable or disable metrics and automated checks to ensure metrics that do not apply to a user's template are included 

![image](https://user-images.githubusercontent.com/78058716/179066062-8176b8c0-f1f7-4f8a-a291-519715120590.png)

### Tests
* Passed Intellij unit tests
* Added integration test for configuration option "EnableErrorMetrics"
* Manual testing was done for each error metric, the function and "do_compute_nodes_have_custom_actions"


### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Added the first version of custom metric filters for default cloudwatch dashboard

added integration tests for new configuration option enable_error_metrics (default true)

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -39,6 +42,10 @@ def new_pcluster_metric(title=None, metrics=None, supported_vol_types=None):
     return _PclusterMetric(title, metrics, supported_vol_types)
 
 
+def custom_metric_filter(error_information=None, metric_name=None, filter_pattern=None):
```
Can we remove the `=None` defaults?

EddyMM replied:
Since this function isn't performing any logic, can we use `_CustomMetricFilter` directly?  Especially since its usages are in the same file.

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -10,6 +10,8 @@
 # limitations under the License.
 from collections import defaultdict, namedtuple
 
+import aws_cdk.aws_cloudwatch
```
Remove this line
There is already:  `from aws_cdk import aws_cloudwatch as cloudwatch`

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -10,6 +10,8 @@
 from collections import defaultdict, namedtuple
 
+import aws_cdk.aws_cloudwatch
+import aws_cdk.aws_logs
```
Change it to `from aws_cdk import aws_logs as logs`

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -51,6 +58,7 @@ def __init__(
         head_node_instance: ec2.CfnInstance,
         shared_storage_infos: dict,
         cw_log_group_name: str,
+        cw_log_group: aws_cdk.aws_logs.CfnLogGroup,
```
Change it to `logs.CfnLogGroup,`

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +230,128 @@ def _add_conditional_storage_widgets(
+
+    def add_custom_error_metrics(self):
+        """Create custom error metric filter and outputs to cloudwatch graph."""
+        jobs_not_starting_errors = [
```
Any reason to make the error_information different from metric name?

HWong-sudo replied:
its like the subcases of the common errors in the intern doc

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +230,128 @@ def _add_conditional_storage_widgets(
+        }
+        self._add_text_widget("## Metrics for Common Errors")
+        widgets_list = []
+        for error_group in error_metric_dict:
```
`for title, metric_filters  in error_metric_dict.items():`

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +230,128 @@ def _add_conditional_storage_widgets(
+        widgets_list = []
+        for error_group in error_metric_dict:
+            metric_list = []
+            for new_filter in error_metric_dict[error_group]:
```
After addressing the comment above, you can use `for metric_filter in metric_filters` directly

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -390,6 +523,15 @@ def _add_cw_log(self):
+                [
+                    self._new_cw_log_widget(
+                        title="Error_Log_Entries",
+                        filters=[self._new_filter(pattern=f"{head_private_ip}.*ERROR")],
```
Why only get errors from head node?

HWong-sudo replied:
oh i forgot to delete that

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -408,6 +550,11 @@ def _add_cw_log(self):
+                    self._new_cw_log_widget(
+                        title="slurm_error",
+                        conditions=[Condition(["slurm"], scheduler)],
+                        filters=[self._new_filter(pattern=f"{head_private_ip}.*ERROR")],
```
Why only get errors from head node?

hanwen-pcluste reviewed (changes requested):
Overall, great work! Thank you!

As we discussed, the new changes need to cover by some unit tests and/or integration tests. We will discuss more tomorrow.

EddyMM commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +230,128 @@ def _add_conditional_storage_widgets(
+                metric_name="Slurm Health Check Failure",
+                filter_pattern="Performing actions for health check type scheduled_events_check",
+            ),
+        ]
```
Can we move this list of `custom_metric_filter`/Metric Filter Patterns from the function to perhaps a top level static variable?
This is just to reduce the amount of code & logic in this function.

HWong-sudo replied:
Done!

EddyMM commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +230,128 @@ def _add_conditional_storage_widgets(
                 widgets_list.append(graph_widget)
         return widgets_list
 
+    def custom_pcluster_metric(self, metric_name, filter_pattern):
```
`custom_pcluster_metric_filter` suits it better since we're creating and returning a MetricFilter

EddyMM commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +230,128 @@ def _add_conditional_storage_widgets(
 
+    def custom_pcluster_metric(self, metric_name, filter_pattern):
+        """Adding custom metric filter from named tuple."""
+        x = aws_cdk.aws_logs.CfnMetricFilter(
```
NIT: `metric_filter` instead of `x`

EddyMM commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +230,128 @@ def _add_conditional_storage_widgets(
+                    metric_name=new_filter.metric_name, filter_pattern=new_filter.filter_pattern
+                )
+                cloudwatch_metric = cloudwatch.Metric(
+                    namespace="ParallelCluster/Errors", metric_name=new_filter.metric_name
```
`ParallelCluster/Errors` has been used in other places, you can move it to a static variable with higher scope.

hanwen-pcluste replied:
Moreover, the name should be related to the cluster name. Otherwise, the metrics will contain data from different clusters

EddyMM replied:
[CloudWatch Dimensions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Dimension) would come in handy here.

hanwen-pcluste replied:
Using Dimensions enables us to share metrics across clusters, saving costs.

EddyMM commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -39,6 +42,10 @@ def new_pcluster_metric(title=None, metrics=None, supported_vol_types=None):
 
 
+def custom_metric_filter(error_information=None, metric_name=None, filter_pattern=None):
+    return _CustomMetricFilter(error_information, metric_name, filter_pattern)
```
Since this function isn't performing any logic, can we use `_CustomMetricFilter` directly?  Especially since its usages are in the same file.

hanwen-pcluste replied:
Done

EddyMM reviewed (commented):
Cool stuff, left some suggestions on improving the readability of the function logic
I'll give some more thought about the namespace being used.

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +228,131 @@ def _add_conditional_storage_widgets(
+            custom_metric_filter(
+                error_information="Script Timeout",
+                metric_name="Infinite Custom Script",
+                filter_pattern="systemd udevd worker timeout kill it",
```
Are you sure the filter pattern is correct?

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +228,131 @@ def _add_conditional_storage_widgets(
+        }
+
+        if self.config.head_node.custom_actions or self.config.scheduling.queues.custom_actions:
+            error_metric_dict["Custom Script Errors"].append(custom_script_errors)
```
Have you tested your code with custom scripts?
I believe the code here should be `error_metric_dict["Custom Script Errors"]=(custom_script_errors)`

HWong-sudo replied:
yeah i changed it to   error_metric_dict.update({"Custom Script Errors": custom_script_errors})

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +228,131 @@ def _add_conditional_storage_widgets(
+    def add_custom_error_metrics(self):
+        """Create custom error metric filter and outputs to cloudwatch graph."""
+        jobs_not_starting_errors = [
+            custom_metric_filter(
+                error_information="Missing IAM Permissions",
+                metric_name="No IAM Group",
+                filter_pattern="UnauthorizedOperation",
+            ),
+            custom_metric_filter(
+                error_information="Missing IAM Permissions",
+                metric_name="Mismatch between IAM Group",
+                filter_pattern="AccessDeniedException",
+            ),
```
What is the difference between the two metrics? How do I reproduce each of them?

Please consider merging them:
```
custom_metric_filter(
                error_information="Missing IAM Permissions",
                metric_name="Missing IAM Permissions",
                filter_pattern="UnauthorizedOperation|AccessDeniedException",
            ),
```

HWong-sudo replied:
techniocally its different, the above one is if the head node doesn't have an IAM group, the bottom one is if you assign it to a different cluster's IAM

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +228,131 @@ def _add_conditional_storage_widgets(
+            custom_metric_filter(
+                error_information="EC2 termination of compute nodes out of PCluster control",
+                metric_name="Terminated EC2 compute node before job submission ",
+                filter_pattern="WARNING Node state check no corresponding instance in EC2 for node",
```
Are you sure this can get the error?
I believe you are missing a `-` after `WARNING`

HWong-sudo replied:
i get an error if i include an alphanumeric character

hanwen-pcluste replied:
Without including the `-` does the metric filter still work?

hanwen-pcluste replied:
You can write something like:
```
r"WARNING \- Node state check no corresponding instance in EC2 for node"
```

hanwen-pcluste replied:
You are right. The syntax of filter pattern is different from Python Regex.
According to the doc https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html, your pattern should work

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +228,131 @@ def _add_conditional_storage_widgets(
+            custom_metric_filter(
+                error_information="Instances failing EC2 health check",
+                metric_name="EC2 Health Check",
+                filter_pattern="WARNING Node state check no corresponding instance in EC2 for node",
```
missing a `-` after `WARNING`

EddyMM commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +228,133 @@ def _add_conditional_storage_widgets(
+            log_group_name=self.cw_log_group_name,
+            metric_transformations=[
+                logs.CfnMetricFilter.MetricTransformationProperty(
+                    metric_namespace=custom_namespace,
```
Instead of having a new namespace for every cluster created, how about using [dimensions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Dimension)?
Dimensions are characteristics associated with a metric (like Metadata). In this  case, a ParallelCluster/Error metric can be  characterised by the cluster, the partition, etc.

Furthermore, if we use a unique namespace for each cluster, the user experience while browsing through the CloudWatch Metrics Page would be as follows for each cluster
```
ParallelCluster/Errors/ClusterA >>> Metrics with no dimensions >>> Metrics
...
```

We could instead publish the metrics to `ParallelCluster/Errors` namespace, while setting the cluster as one of the Metric Dimensions e.g.
```
metric_transformations=[
    logs.CfnMetricFilter.MetricTransformationProperty(
    	...
    	dimensions=[
    		logs.CfnMetricFilter.DimensionProperty(
    			key="cluster",
    			value=self.config.cluster_name
    		)
    	],
    	...
```

Such that the user will see/use one namespace i.e. `ParallelCluster/Errors` which can then be classified by Dimension i.e. cluster. The experience would then be:

```
ParallelCluster/Errors >>> By Cluster >>> Metrics
```

---
It's similar to how EC2 has one namespace but you can view metrics "By InstanceId"

hanwen-pcluste replied:
Thank you for the suggestion. @HWong-sudo, can you change the code?

HWong-sudo replied:
encountered issues with adding dimensions. When copied from documentation, got message": "type object 'CfnMetricFilter' has no attribute 'DimensionProperty. MetricTransformation property also does not have dimension attribute within my local init. Plus in this cloudformation PR (https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-logs/issues/64) they mention: 

"_The following announcement makes dimensions available for metric filters.

https://aws.amazon.com/about-aws/whats-new/2021/05/amazon-cloudwatch-logs-announces-dimension-support-for-metric-filters/

However, it appears to be missing **when specifying the resource through CloudFormation**. Getting the following error on stack events console.

Encountered unsupported property Dimensions
The Python SDK documentation correctly shows this as available, but missing in CloudFormation docs for CloudWatchLogs.
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs.html#CloudWatchLogs.Client.put_metric_filter

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-logs-metricfilter.html_"

They mention that on 7/16/2022 ( https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-logs/issues/64#issuecomment-1157913657) that the feature "_has issues related to deployment and traffic migration_" and the latest update on 7/25/2022 says that the "_feature is not available on eu-central-1_" There has been no further update as of 7/29/2022

Based on the above evidence, I conclude that the dimensions are not available to use as for CFN metricFilter's metric transformation property. Thus I will not be implementing this feature. It can be postponed to a future goal

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +229,116 @@ def _add_conditional_storage_widgets(
+        jobs_not_starting_errors = [
+            custom_metric_filter(
+                metric_name="Mismatch between IAM Group",
+                filter_pattern="UnauthorizedOperation|AccessDeniedException",
```
My bad. The syntax here is wrong

According to https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html, it should be "?UnauthorizedOperation ?AccessDeniedException"

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -135,6 +144,7 @@ def _add_resources(self):
         # Head Node logs, if CW Logs are enabled
```
The comment should be improved

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +229,116 @@ def _add_conditional_storage_widgets(
+        metric_filter.add_depends_on(self.cw_log_group)
+        return metric_filter
+
+    def add_custom_error_metrics(self):
```
Shall we add a general error metric?
```
custom_metric_filter(
                metric_name="All errors",
                filter_pattern="?error ?Error ?ERROR",
            )
```

HWong-sudo replied:
I was talking to maurizo and that might be useful, wonder where to add it tho, also prob could add WARNING

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -39,6 +42,10 @@ def new_pcluster_metric(title=None, metrics=None, supported_vol_types=None):
     return _PclusterMetric(title, metrics, supported_vol_types)
 
 
+def custom_metric_filter( metric_name=None, filter_pattern=None):
```
Copy Eddy's comment here:
> Since this function isn't performing any logic, can we use _CustomMetricFilter directly? Especially since its usages are in the same file.

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +229,116 @@ def _add_conditional_storage_widgets(
+                logs.CfnMetricFilter.MetricTransformationProperty(
+                    metric_namespace=custom_namespace,
+                    metric_name=metric_name,
+                    metric_value="100",
```
This should be
```
metric_value="1",
```

HWong-sudo replied:
yeah but the value is sometimes too small, value still not even 1 in the actual graph

hanwen-pcluste commented on cli/src/pcluster/templates/cw_dashboard_builder.py:
```diff
@@ -219,6 +229,122 @@ def _add_conditional_storage_widgets(
+        custom_script_errors = [
+            _CustomMetricFilter(
+                metric_name="Cannot retrieve custom script",
+                filter_pattern="error occurred .* when calling the HeadObject operation Forbidden",
```
My bad.
According to https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html, it should be:
"error occurred when calling the HeadObject operation"

Commit:
added configuration option enable_error_metrics (default true)

hanwen-pcluste commented on cli/src/pcluster/resources/compute_node/user_data.sh:
```diff
@@ -114,7 +114,8 @@ write_files:
       function error_exit
       {
-        echo "Bootstrap failed with error: $1"
+        echo "B]\
+        ootstrap failed with error: $1"
```
Typo

Commit:
added integration tests for new configuration option enable_error_metrics (default true)

added integration tests for new configuration option enable_error_metrics (default true)

added integration tests for new configuration option enable_error_metrics (default true)

hanwen-pcluste commented on tests/integration-tests/tests/dashboard/test_dashboard.py:
```diff
@@ -38,6 +41,10 @@ def test_dashboard(
+                assert_that(response["DashboardBody"]).contains("Metrics for Common Errors")
+            else:
+                assert_that(response["DashboardBody"]).does_not_contain("Metrics for Common Errors")
         else:
             assert_that(response["DashboardBody"]).does_not_contain("Head Node Logs")
```
Can you assert the graph does not exist here?
